File: leetcode/go/2919.minimum-increment-operations-to-make-array-beautiful.go
func minIncrementOperations(nums []int, k int) int64 {
    var dp, p1, p2, p3 int64 = 0, 0, 0, 0
    var k64 int64 = int64(k)
    for _, n := range nums {
        dp = min(p1, min(p2, p3)) + max(0, k64-int64(n))
        p3 = p2
        p2 = p1
        p1 = dp
    }
    return min(p1, min(p2, p3))
}

File: leetcode/rust-leetcode/src/2919.minimum-increment-operations-to-make-array-beautiful.rs
use std::cmp::{min,max};
impl Solution {
    pub fn min_increment_operations(nums: Vec<i32>, k: i32) -> i64 {
        let k = k as i64;
        let nums = nums.iter().map(|&n| n as i64).collect::<Vec<i64>>();
        let (mut p1, mut p2, mut p3, mut dp) = (0, 0, 0, 0);
        for n in nums {
            dp = min(p1, min(p2, p3)) + max(0, k - n);
            p3 = p2;
            p2 = p1;
            p1 = dp;
        }
        min(p1, min(p2, p3))
    }
}

